internal/querynodev2/segments/manager.go

```diff
@@ -388,64 +385,116 @@ func (mgr *segmentManager) Empty() bool {
 	switch scope {
 	case querypb.DataScope_Streaming:
-		if remove(segmentID, mgr.growingSegments) {
+		growing = mgr.removeSegmentWithType(SegmentTypeGrowing, segmentID)
+		if growing != nil {
 			removeGrowing = 1
 		}
 
 	case querypb.DataScope_Historical:
-		if remove(segmentID, mgr.sealedSegments) {
+		sealed = mgr.removeSegmentWithType(SegmentTypeSealed, segmentID)
+		if sealed != nil {
 			removeSealed = 1
 		}
 
 	case querypb.DataScope_All:
-		if remove(segmentID, mgr.growingSegments) {
+		growing = mgr.removeSegmentWithType(SegmentTypeGrowing, segmentID)
+		if growing != nil {
 			removeGrowing = 1
 		}
-		if remove(segmentID, mgr.sealedSegments) {
+
+		sealed = mgr.removeSegmentWithType(SegmentTypeSealed, segmentID)
+		if sealed != nil {
 			removeSealed = 1
 		}
 	}
-
 	mgr.updateMetric()
+	mgr.mu.Unlock()
+
+	if growing != nil {
+		remove(growing)
+	}
+
+	if sealed != nil {
+		remove(sealed)
+	}
+
 	return removeGrowing, removeSealed
 }
 
+func (mgr *segmentManager) removeSegmentWithType(typ SegmentType, segmentID UniqueID) *LocalSegment {
+	switch typ {
+	case SegmentTypeGrowing:
+		s, ok := mgr.growingSegments[segmentID]
+		if ok {
+			delete(mgr.growingSegments, segmentID)
+			return s.(*LocalSegment)
+		}
+
+	case SegmentTypeSealed:
+		s, ok := mgr.sealedSegments[segmentID]
+		if ok {
+			delete(mgr.sealedSegments, segmentID)
+			return s.(*LocalSegment)
+		}
+	default:
+		return nil
+	}
+
+	return nil
+}
+
 func (mgr *segmentManager) RemoveBy(filters ...SegmentFilter) (int, int) {
 	mgr.mu.Lock()
-	defer mgr.mu.Unlock()
 
-	var removeGrowing, removeSealed int
+	var removeGrowing, removeSealed []*LocalSegment
 	for id, segment := range mgr.growingSegments {
-		if filter(segment, filters...) && remove(id, mgr.growingSegments) {
-			removeGrowing++
+		if filter(segment, filters...) {
+			s := mgr.removeSegmentWithType(SegmentTypeGrowing, id)
+			if s != nil {
+				removeGrowing = append(removeGrowing, s)
+			}
 		}
 	}
 
 	for id, segment := range mgr.sealedSegments {
-		if filter(segment, filters...) && remove(id, mgr.sealedSegments) {
-			removeSealed++
+		if filter(segment, filters...) {
+			s := mgr.removeSegmentWithType(SegmentTypeSealed, id)
+			if s != nil {
+				removeSealed = append(removeSealed, s)
+			}
 		}
 	}
-
 	mgr.updateMetric()
-	return removeGrowing, removeSealed
+	mgr.mu.Unlock()
+
+	for _, s := range removeGrowing {
+		remove(s)
+	}
+
+	for _, s := range removeSealed {
+		remove(s)
+	}
+
+	return len(removeGrowing), len(removeSealed)
 }
 
 func (mgr *segmentManager) Clear() {
 	mgr.mu.Lock()
 	defer mgr.mu.Unlock()
 
-	for id := range mgr.growingSegments {
-		remove(id, mgr.growingSegments)
+	for id, segment := range mgr.growingSegments {
+		delete(mgr.growingSegments, id)
+		remove(segment.(*LocalSegment))
 	}
 
-	for id := range mgr.sealedSegments {
-		remove(id, mgr.sealedSegments)
+	for id, segment := range mgr.sealedSegments {
+		delete(mgr.sealedSegments, id)
+		remove(segment.(*LocalSegment))
```

Comment:
if `Clear` & `Unpin` happens concurrently, this will always cause deadlock here

Reply:
Unpin won't acquire  manager's lock here